packages/config-loader/src/schema/compile.ts

```diff
@@ -101,17 +101,33 @@ export function compileConfigSchemas(
 
   const merged = mergeConfigSchemas(schemas.map(_ => _.value));
 
-  if (options?.noUndeclaredProperties) {
-    traverse(merged, (schema: SchemaObject) => {
-      /**
-       * The `additionalProperties` key can only be applied to `type: object` in the JSON
-       *  schema.
-       */
-      if (schema?.type === 'object') {
-        schema.additionalProperties ||= false;
+  traverse(
+    merged,
+    (
+      schema: SchemaObject,
+      _: string,
+      _1?: any,
+      _2?: any,
+      _3?: any,
+      parentSchema?: SchemaObject,
+    ) => {
+      // This runs before we compile the AJV so visibilityByDataPath has the
+      //  correct data.
+      if (parentSchema?.visibility === 'secret') {
+        schema.visibility = 'secret';
```

Comment:
I feel that doing this unconditionally is a bit tricky, since it'll override any deeper visibility configs. For example, this wouldn't really behave as one would expect:

```ts
interface Config {
  /** @visibility secret */
  foo: {
    /** @visibility frontend */
    bar: string
  }
}
```

We could make explicitly defined visibility interrupt this chaining, as in switching this over to `??=`.

Another option is to add a new way to explicitly define a deep visibility. Considering the current use-case in isolation that might not make as much sense, but we have actually discovered that we'll likely need something similar for frontend config too. For example we could consider adding something like this:

```ts
interface Config {
  /** @deepVisibility secret */
  foo: { ... }
}
```

What do you think?

Reply:
A `@deepVisibility` tag I think would have the same transitive effect as the one you implement in this PR. On top of that we can then check whether there's a mismatch between the declared deep visibility and direct visibility in the same way as we already validate the merged visibility.

Reply:
Ah yup, that makes sense. I'll take a stab at it tonight.